File: android/app/src/test/java/ru/cloudly/sync/MirrorRulesTest.kt
package ru.cloudly.sync

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import ru.cloudly.sync.mirror.LocalDir
import ru.cloudly.sync.mirror.LocalFile
import ru.cloudly.sync.mirror.LocalSnapshot
import ru.cloudly.sync.mirror.MirrorRow
import ru.cloudly.sync.mirror.MirrorRules

/**
 * Сверка зеркала: ошибка здесь означает либо потерянный файл, либо удаление чужого содержимого
 * в облаке. Обе крайности дорогие, поэтому правила проверяются без устройства.
 */
class MirrorRulesTest {

    private val now = 1_000_000L

    private fun local(
        path: String = "/s/Download/doc.pdf",
        size: Long = 100,
        mtime: Long = now - 60_000,
        inode: Long = 42,
    ) = LocalFile(
        path = path,
        name = path.substringAfterLast('/'),
        dir = path.substringBeforeLast('/'),
        relDir = "Download",
        root = "/s/Download",
        size = size,
        mtime = mtime,
        inode = inode,
    )

    private fun row(
        path: String = "/s/Download/doc.pdf",
        entryId: String = "entry-1",
        size: Long = 100,
        mtime: Long = now - 60_000,
        inode: Long = 42,
        sha256: String? = "aa",
    ) = MirrorRow(path, "cloud-dir", entryId, inode, size, mtime, sha256)

    private fun snapshot(files: List<LocalFile>, unreadable: Int = 0, capped: Boolean = false) =
        LocalSnapshot(files, listOf(LocalDir("/s/Download", "Download")), unreadable, capped)

    @Test
    fun newFileIsUploaded() {
        val plan = MirrorRules.plan(listOf(local()), emptyMap(), now, deletionsAllowed = true)
        assertEquals(listOf("/s/Download/doc.pdf"), plan.uploads.map { it.path })
        assertTrue(plan.renames.isEmpty())
        assertTrue(plan.deletes.isEmpty())
    }

    @Test
    fun fileStillBeingWrittenWaits() {
        // файл изменился только что: его ещё пишут, выгружать нельзя
        val fresh = local(mtime = now - 1_000)
        val plan = MirrorRules.plan(listOf(fresh), emptyMap(), now, deletionsAllowed = true)
        assertTrue(plan.uploads.isEmpty())
    }

    @Test
    fun unchangedFileIsNotUploadedAgain() {
        val plan = MirrorRules.plan(listOf(local()), mapOf("/s/Download/doc.pdf" to row()), now, deletionsAllowed = true)
        assertTrue(plan.uploads.isEmpty())
        assertTrue(plan.deletes.isEmpty())
    }

    @Test
    fun changedContentIsUploadedAgain() {
        val edited = local(size = 200)
        val plan = MirrorRules.plan(listOf(edited), mapOf("/s/Download/doc.pdf" to row()), now, deletionsAllowed = true)
        assertEquals(listOf("/s/Download/doc.pdf"), plan.uploads.map { it.path })
        assertTrue(plan.deletes.isEmpty())
    }

    @Test
    fun renameIsNotDeleteAndUpload() {
        // тот же inode, другой путь: байты передавать не нужно, содержимое в облаке уже есть
        val moved = local(path = "/s/Download/отчёт.pdf", inode = 42)
        val plan = MirrorRules.plan(listOf(moved), mapOf("/s/Download/doc.pdf" to row()), now, deletionsAllowed = true)
        assertEquals(1, plan.renames.size)
        assertEquals("/s/Download/doc.pdf", plan.renames.first().first.path)
        assertEquals("/s/Download/отчёт.pdf", plan.renames.first().second.path)
        assertTrue(plan.uploads.isEmpty())
        assertTrue(plan.deletes.isEmpty())
    }

    @Test
    fun renameWithEditIsNotRename() {
        // переименование и правка разом: размер разошёлся, значит это не переименование,
        // а «пропал старый + появился новый» — так облачное содержимое не перезаписывается
        val moved = local(path = "/s/Download/отчёт.pdf", size = 200, inode = 42)
        val plan = MirrorRules.plan(listOf(moved), mapOf("/s/Download/doc.pdf" to row()), now, deletionsAllowed = true)
        assertTrue(plan.renames.isEmpty())
        assertEquals(listOf("/s/Download/отчёт.pdf"), plan.uploads.map { it.path })
        assertEquals(listOf("entry-1"), plan.deletes.map { it.entryId })
    }

    @Test
    fun reusedInodeWithOtherDateIsNotRename() {
        // ядро отдало освободившийся inode новому файлу: размер совпал, дата нет — не переименование
        val fresh = local(path = "/s/Download/new.pdf", size = 100, mtime = now - 5_000, inode = 42)
        val plan = MirrorRules.plan(listOf(fresh), mapOf("/s/Download/doc.pdf" to row()), now, deletionsAllowed = true)
        assertTrue(plan.renames.isEmpty())
        assertEquals(listOf("entry-1"), plan.deletes.map { it.entryId })
    }

    @Test
    fun hiddenAndJunkRowsAreNeverDeletedInCloud() {
        // обход такие имена не показывает: «файла нет в снимке» — это правило показа, а не удаление
        val known = mapOf(
            "/s/Download/.nomedia" to row(path = "/s/Download/.nomedia", entryId = "hidden"),
            "/s/Download/movie.mp4.part" to row(path = "/s/Download/movie.mp4.part", entryId = "junk"),
            "/s/Download/doc.pdf" to row(path = "/s/Download/doc.pdf", entryId = "real"),
        )
        val plan = MirrorRules.plan(listOf(local()), known, now, deletionsAllowed = true)
        assertTrue(plan.deletes.none { it.entryId == "hidden" || it.entryId == "junk" })
        assertTrue(plan.deletes.isEmpty())
    }

    @Test
    fun futureMtimeIsStable() {
        // файл из архива с датой в будущем не должен застрять навсегда
        assertTrue(MirrorRules.isStable(now + 5_000, now))
    }

    @Test
    fun missingFileIsDeletedInCloud() {
        val plan = MirrorRules.plan(emptyList(), mapOf("/s/Download/doc.pdf" to row()), now, deletionsAllowed = true)
        assertEquals(listOf("entry-1"), plan.deletes.map { it.entryId })
        assertFalse(plan.blocked)
    }

    @Test
    fun unknownInodeFallsBackToDeleteAndUpload() {
        // inode узнать не удалось: переименование неотличимо от «удалили и создали»
        val fresh = local(path = "/s/Download/отчёт.pdf", inode = 0)
        val plan = MirrorRules.plan(listOf(fresh), mapOf("/s/Download/doc.pdf" to row(inode = 0)), now, deletionsAllowed = true)
        assertTrue(plan.renames.isEmpty())
        assertEquals(listOf("/s/Download/отчёт.pdf"), plan.uploads.map { it.path })
        assertEquals(listOf("entry-1"), plan.deletes.map { it.entryId })
    }

    @Test
    fun unreadableFolderBlocksDeletions() {
        // папка не читается — «файла нет» означает «мы его не увидели», а не «его удалили»
        val snap = snapshot(emptyList(), unreadable = 3)
        assertFalse(MirrorRules.deletionsAllowed(snap))
        val plan = MirrorRules.plan(snap.files, mapOf("/s/Download/doc.pdf" to row()), now, deletionsAllowed = false)
        assertTrue(plan.deletes.isEmpty())
        assertTrue(plan.blocked)
        assertEquals(1, plan.blockedCount)
    }

    @Test
    fun incompleteScanBlocksDeletions() {
        assertFalse(MirrorRules.deletionsAllowed(snapshot(emptyList(), capped = true)))
    }

    @Test
    fun massDeleteNeedsConfirmation() {
        // пропало 30 из 100 — четверть и больше порога: удалять нельзя без подтверждения
        assertTrue(MirrorRules.massDelete(30, 100))
        val known = (1..100).associate { "/s/Download/f$it" to row(path = "/s/Download/f$it") }
        val plan = MirrorRules.plan(emptyList(), known, now, deletionsAllowed = true)
        assertTrue(plan.deletes.isEmpty())
        assertTrue(plan.blocked)
        assertEquals(100, plan.blockedCount)
    }

    @Test
    fun smallShareOfLargeLibraryIsFine() {
        // 25 файлов из 1000 — обычное дело (чистка загрузок), предохранитель не мешает
        assertFalse(MirrorRules.massDelete(25, 1000))
    }

    @Test
    fun hugeDeletionIsBlockedEvenInSmallLibrary() {
        assertTrue(MirrorRules.massDelete(MirrorRules.MASS_DELETE_MAX, 100_000))
    }

    @Test
    fun confirmedDeletionGoesThrough() {
        val known = (1..100).associate { "/s/Download/f$it" to row(path = "/s/Download/f$it", entryId = "e$it") }
        val plan = MirrorRules.plan(emptyList(), known, now, deletionsAllowed = true, confirmed = true)
        assertEquals(100, plan.deletes.size)
        assertFalse(plan.blocked)
    }

    @Test
    fun emptyLibraryDoesNotTripTheGuard() {
        assertFalse(MirrorRules.massDelete(0, 0))
        val plan = MirrorRules.plan(emptyList(), emptyMap(), now, deletionsAllowed = true)
        assertTrue(plan.deletes.isEmpty())
        assertFalse(plan.blocked)
    }

    @Test
    fun conflictCopyKeepsBothVersions() {
        val name = MirrorRules.conflictName("отчёт.pdf", 1_700_000_000_000L)
        assertTrue(name.startsWith("отчёт (конфликт "))
        assertTrue(name.endsWith(".pdf"))
        // файл без расширения тоже получает осмысленное имя
        assertTrue(MirrorRules.conflictName("README", 1_700_000_000_000L).startsWith("README (конфликт "))
    }

    @Test
    fun rowsOfUnselectedFolderAreLeftAlone() {
        // папку сняли с выбора: её файлов в снимке нет, и удалять их в облаке нельзя —
        // пользователь всего лишь снял галочку, а не удалил данные
        val known = mapOf(
            "/s/Download/doc.pdf" to row(path = "/s/Download/doc.pdf", entryId = "in-mirror"),
            "/s/DCIM/old.jpg" to row(path = "/s/DCIM/old.jpg", entryId = "not-selected"),
        )
        val inRoots = MirrorRules.underRoots(known, listOf("/s/Download"))
        assertEquals(setOf("/s/Download/doc.pdf"), inRoots.keys)
        val plan = MirrorRules.plan(emptyList(), inRoots, now, deletionsAllowed = true)
        assertEquals(listOf("in-mirror"), plan.deletes.map { it.entryId })
    }

    @Test
    fun stabilityWindow() {
        assertTrue(MirrorRules.isStable(now - MirrorRules.STABLE_MS, now))
        assertFalse(MirrorRules.isStable(now - MirrorRules.STABLE_MS + 1, now))
    }
}

File: flutter/test/sync/mirror_rules_test.dart
import 'package:cloudly_flutter/sync/mirror/mirror_models.dart';
import 'package:cloudly_flutter/sync/mirror/mirror_rules.dart';
import 'package:flutter_test/flutter_test.dart';

/// Сверка зеркала: ошибка здесь означает либо потерянный файл, либо удаление чужого содержимого
/// в облаке. Обе крайности дорогие, поэтому правила проверяются без устройства.
void main() {
  const now = 1000000;

  String afterLastSlash(String path) =>
      path.substring(path.lastIndexOf('/') + 1);

  String beforeLastSlash(String path) =>
      path.substring(0, path.lastIndexOf('/'));

  LocalFile local({
    String path = '/s/Download/doc.pdf',
    int size = 100,
    int mtime = now - 60000,
    int inode = 42,
  }) => LocalFile(
    path: path,
    name: afterLastSlash(path),
    dir: beforeLastSlash(path),
    relDir: 'Download',
    root: '/s/Download',
    size: size,
    mtime: mtime,
    inode: inode,
  );

  MirrorRow row({
    String path = '/s/Download/doc.pdf',
    String entryId = 'entry-1',
    int size = 100,
    int mtime = now - 60000,
    int inode = 42,
    String? sha256 = 'aa',
  }) => MirrorRow(
    path: path,
    cloudFolderId: 'cloud-dir',
    entryId: entryId,
    inode: inode,
    size: size,
    mtime: mtime,
    sha256: sha256,
  );

  LocalSnapshot snapshot(
    List<LocalFile> files, {
    int unreadable = 0,
    bool capped = false,
  }) => LocalSnapshot(
    files: files,
    dirs: const [LocalDir('/s/Download', 'Download')],
    unreadable: unreadable,
    capped: capped,
  );

  group('MirrorRules', () {
    test('новый файл выгружается', () {
      final plan = MirrorRules.plan(
        local: [local()],
        known: const {},
        now: now,
        deletionsAllowed: true,
      );

      expect(plan.uploads.map((f) => f.path), ['/s/Download/doc.pdf']);
      expect(plan.renames, isEmpty);
      expect(plan.deletes, isEmpty);
    });

    test('файл, который ещё пишут, ждёт', () {
      // файл изменился только что: его ещё пишут, выгружать нельзя
      final fresh = local(mtime: now - 1000);
      final plan = MirrorRules.plan(
        local: [fresh],
        known: const {},
        now: now,
        deletionsAllowed: true,
      );

      expect(plan.uploads, isEmpty);
    });

    test('неизменившийся файл не выгружается повторно', () {
      final plan = MirrorRules.plan(
        local: [local()],
        known: {'/s/Download/doc.pdf': row()},
        now: now,
        deletionsAllowed: true,
      );

      expect(plan.uploads, isEmpty);
      expect(plan.deletes, isEmpty);
    });

    test('изменившееся содержимое выгружается снова', () {
      final edited = local(size: 200);
      final plan = MirrorRules.plan(
        local: [edited],
        known: {'/s/Download/doc.pdf': row()},
        now: now,
        deletionsAllowed: true,
      );

      expect(plan.uploads.map((f) => f.path), ['/s/Download/doc.pdf']);
      expect(plan.deletes, isEmpty);
    });

    test('переименование — это не удаление с выгрузкой', () {
      // тот же inode, другой путь: байты передавать не нужно, содержимое в облаке уже есть
      final moved = local(path: '/s/Download/отчёт.pdf', inode: 42);
      final plan = MirrorRules.plan(
        local: [moved],
        known: {'/s/Download/doc.pdf': row()},
        now: now,
        deletionsAllowed: true,
      );

      expect(plan.renames.length, 1);
      expect(plan.renames.first.$1.path, '/s/Download/doc.pdf');
      expect(plan.renames.first.$2.path, '/s/Download/отчёт.pdf');
      expect(plan.uploads, isEmpty);
      expect(plan.deletes, isEmpty);
    });

    test('переименование вместе с правкой — не переименование', () {
      // переименование и правка разом: размер разошёлся, значит это не переименование,
      // а «пропал старый + появился новый» — так облачное содержимое не перезаписывается
      final moved = local(path: '/s/Download/отчёт.pdf', size: 200, inode: 42);
      final plan = MirrorRules.plan(
        local: [moved],
        known: {'/s/Download/doc.pdf': row()},
        now: now,
        deletionsAllowed: true,
      );

      expect(plan.renames, isEmpty);
      expect(plan.uploads.map((f) => f.path), ['/s/Download/отчёт.pdf']);
      expect(plan.deletes.map((r) => r.entryId), ['entry-1']);
    });

    test('переиспользованный inode с другой датой — не переименование', () {
      // ядро отдало освободившийся inode новому файлу: размер совпал, дата нет — не переименование
      final fresh = local(
        path: '/s/Download/new.pdf',
        size: 100,
        mtime: now - 5000,
        inode: 42,
      );
      final plan = MirrorRules.plan(
        local: [fresh],
        known: {'/s/Download/doc.pdf': row()},
        now: now,
        deletionsAllowed: true,
      );

      expect(plan.renames, isEmpty);
      expect(plan.deletes.map((r) => r.entryId), ['entry-1']);
    });

    test('скрытые и служебные строки в облаке не удаляются никогда', () {
      // обход такие имена не показывает: «файла нет в снимке» — это правило показа, а не удаление
      final known = {
        '/s/Download/.nomedia': row(
          path: '/s/Download/.nomedia',
          entryId: 'hidden',
        ),
        '/s/Download/movie.mp4.part': row(
          path: '/s/Download/movie.mp4.part',
          entryId: 'junk',
        ),
        '/s/Download/doc.pdf': row(
          path: '/s/Download/doc.pdf',
          entryId: 'real',
        ),
      };
      final plan = MirrorRules.plan(
        local: [local()],
        known: known,
        now: now,
        deletionsAllowed: true,
      );

      expect(
        plan.deletes.where((r) => r.entryId == 'hidden' || r.entryId == 'junk'),
        isEmpty,
      );
      expect(plan.deletes, isEmpty);
    });

    test('дата из будущего считается устоявшейся', () {
      // файл из архива с датой в будущем не должен застрять навсегда
      expect(MirrorRules.isStable(now + 5000, now), isTrue);
    });

    test('пропавший файл удаляется в облаке', () {
      final plan = MirrorRules.plan(
        local: const [],
        known: {'/s/Download/doc.pdf': row()},
        now: now,
        deletionsAllowed: true,
      );

      expect(plan.deletes.map((r) => r.entryId), ['entry-1']);
      expect(plan.blocked, isFalse);
    });

    test('неизвестный inode откатывается к «удалить и выгрузить»', () {
      // inode узнать не удалось: переименование неотличимо от «удалили и создали»
      final fresh = local(path: '/s/Download/отчёт.pdf', inode: 0);
      final plan = MirrorRules.plan(
        local: [fresh],
        known: {'/s/Download/doc.pdf': row(inode: 0)},
        now: now,
        deletionsAllowed: true,
      );

      expect(plan.renames, isEmpty);
      expect(plan.uploads.map((f) => f.path), ['/s/Download/отчёт.pdf']);
      expect(plan.deletes.map((r) => r.entryId), ['entry-1']);
    });

    test('нечитаемая папка запрещает удаления', () {
      // папка не читается — «файла нет» означает «мы его не увидели», а не «его удалили»
      final snap = snapshot(const [], unreadable: 3);

      expect(MirrorRules.deletionsAllowed(snap), isFalse);
      // и обратная сторона: по чистому снимку удаления разрешены — иначе облако
      // никогда бы ничего не убирало, а тесты видели бы только запрет
      expect(MirrorRules.deletionsAllowed(snapshot(const [])), isTrue);

      final plan = MirrorRules.plan(
        local: snap.files,
        known: {'/s/Download/doc.pdf': row()},
        now: now,
        deletionsAllowed: false,
      );

      expect(plan.deletes, isEmpty);
      expect(plan.blocked, isTrue);
      expect(plan.blockedCount, 1);
    });

    test('неполный обход запрещает удаления', () {
      expect(
        MirrorRules.deletionsAllowed(snapshot(const [], capped: true)),
        isFalse,
      );
    });

    test('массовое удаление требует подтверждения', () {
      // пропало 30 из 100 — четверть и больше порога: удалять нельзя без подтверждения
      expect(MirrorRules.massDelete(30, 100), isTrue);

      final known = {
        for (var i = 1; i <= 100; i++)
          '/s/Download/f$i': row(path: '/s/Download/f$i'),
      };
      final plan = MirrorRules.plan(
        local: const [],
        known: known,
        now: now,
        deletionsAllowed: true,
      );

      expect(plan.deletes, isEmpty);
      expect(plan.blocked, isTrue);
      expect(plan.blockedCount, 100);
    });

    test('малая доля от большой библиотеки — это нормально', () {
      // 25 файлов из 1000 — обычное дело (чистка загрузок), предохранитель не мешает
      expect(MirrorRules.massDelete(25, 1000), isFalse);
    });

    test('огромное удаление блокируется даже в маленькой библиотеке', () {
      expect(MirrorRules.massDelete(MirrorRules.massDeleteMax, 100000), isTrue);
    });

    test('подтверждённое удаление проходит', () {
      final known = {
        for (var i = 1; i <= 100; i++)
          '/s/Download/f$i': row(path: '/s/Download/f$i', entryId: 'e$i'),
      };
      final plan = MirrorRules.plan(
        local: const [],
        known: known,
        now: now,
        deletionsAllowed: true,
        confirmed: true,
      );

      expect(plan.deletes.length, 100);
      expect(plan.blocked, isFalse);
    });

    test('пустая библиотека не срабатывает на предохранителе', () {
      expect(MirrorRules.massDelete(0, 0), isFalse);

      final plan = MirrorRules.plan(
        local: const [],
        known: const {},
        now: now,
        deletionsAllowed: true,
      );

      expect(plan.deletes, isEmpty);
      expect(plan.blocked, isFalse);
    });

    test('конфликтная копия сохраняет обе версии', () {
      final name = MirrorRules.conflictName('отчёт.pdf', 1700000000000);

      expect(name.startsWith('отчёт (конфликт '), isTrue);
      expect(name.endsWith('.pdf'), isTrue);
      // файл без расширения тоже получает осмысленное имя
      expect(
        MirrorRules.conflictName(
          'README',
          1700000000000,
        ).startsWith('README (конфликт '),
        isTrue,
      );
    });

    test('метка конфликта держит формат «yyyy-MM-dd HH.mm»', () {
      // формат — часть имени файла в облаке: смена вида ломает узнаваемость копий
      final at = DateTime(2024, 3, 7, 9, 5).millisecondsSinceEpoch;

      expect(
        MirrorRules.conflictName('отчёт.pdf', at),
        'отчёт (конфликт 2024-03-07 09.05).pdf',
      );
      expect(
        MirrorRules.conflictName('README', at),
        'README (конфликт 2024-03-07 09.05)',
      );
    });

    test('строки снятой с выбора папки не трогаются', () {
      // папку сняли с выбора: её файлов в снимке нет, и удалять их в облаке нельзя —
      // пользователь всего лишь снял галочку, а не удалил данные
      final known = {
        '/s/Download/doc.pdf': row(
          path: '/s/Download/doc.pdf',
          entryId: 'in-mirror',
        ),
        '/s/DCIM/old.jpg': row(
          path: '/s/DCIM/old.jpg',
          entryId: 'not-selected',
        ),
      };
      bool inRoots(String path) =>
          MirrorRules.underRoots(path, ['/s/Download']);

      expect(inRoots('/s/Download/doc.pdf'), isTrue);
      expect(inRoots('/s/DCIM/old.jpg'), isFalse);

      final plan = MirrorRules.plan(
        local: const [],
        known: known,
        now: now,
        deletionsAllowed: true,
        inScope: inRoots,
      );

      expect(plan.deletes.map((r) => r.entryId), ['in-mirror']);
    });

    test('окно стабильности', () {
      expect(MirrorRules.isStable(now - MirrorRules.stableMs, now), isTrue);
      expect(
        MirrorRules.isStable(now - MirrorRules.stableMs + 1, now),
        isFalse,
      );
    });

    // ── Дополнено сверх Kotlin-набора ────────────────────────────────────────
    // Проверено мутациями (правило ломается нарочно — тесты обязаны упасть): эти ветки
    // в Kotlin-наборе тоже ничем не прикрыты, а цена ошибки здесь — потерянная правка
    // или дважды переписанная облачная запись.

    test('правка одной даты — тоже изменение', () {
      // размер тот же, дата другая: сверять только размер мало — правка не уехала бы
      final touched = local(mtime: now - 30000);
      final plan = MirrorRules.plan(
        local: [touched],
        known: {'/s/Download/doc.pdf': row(mtime: now - 60000)},
        now: now,
        deletionsAllowed: true,
      );

      expect(plan.uploads.map((f) => f.path), ['/s/Download/doc.pdf']);
    });

    test('один inode не даёт двух переименований', () {
      // два файла с одним inode: переименованием признаём только первое совпадение,
      // иначе одна облачная запись переписывалась бы дважды
      final plan = MirrorRules.plan(
        local: [
          local(path: '/s/Download/a.pdf'),
          local(path: '/s/Download/b.pdf'),
        ],
        known: {'/s/Download/doc.pdf': row()},
        now: now,
        deletionsAllowed: true,
      );

      expect(plan.renames.length, 1);
      expect(plan.renames.first.$2.path, '/s/Download/a.pdf');
      expect(plan.uploads.map((f) => f.path), ['/s/Download/b.pdf']);
      expect(plan.deletes, isEmpty);
    });
  });

  group('MirrorRules.emptyFolderCandidates', () {
    test('папка, которой нет на телефоне, уходит в уборку', () {
      final out = MirrorRules.emptyFolderCandidates(
        dirs: {'cloud-old': '/s/Download/Старое имя'},
        aliveLocally: {'/s/Download/Новое имя'},
        roots: ['/s/Download'],
        rootCloudIds: {'cloud-root'},
      );
      expect(out, ['/s/Download/Старое имя']);
    });

    test('корень зеркала не убирается никогда', () {
      // корень — это адрес, по которому лежит всё зеркало: удалить его значит осиротить папку
      final out = MirrorRules.emptyFolderCandidates(
        dirs: {'cloud-root': '/s/Download'},
        aliveLocally: const {},
        roots: ['/s/Download'],
        rootCloudIds: {'cloud-root'},
      );
      expect(out, isEmpty);
    });

    test('папка, которая на месте, не трогается', () {
      final out = MirrorRules.emptyFolderCandidates(
        dirs: {'cloud-a': '/s/Download/Отчёты'},
        aliveLocally: {'/s/Download/Отчёты'},
        roots: ['/s/Download'],
        rootCloudIds: {'cloud-root'},
      );
      expect(out, isEmpty);
    });

    test('папка снятая с выбора не трогается', () {
      // её нет в снимке только потому, что галочку сняли: удалять в облаке нечего
      final out = MirrorRules.emptyFolderCandidates(
        dirs: {'cloud-b': '/s/Pictures/Старое'},
        aliveLocally: const {},
        roots: ['/s/Download'],
        rootCloudIds: const {},
      );
      expect(out, isEmpty);
    });

    test('глубокие папки идут первыми', () {
      // тогда родитель, опустевший после удаления детей, уходит в том же проходе
      final out = MirrorRules.emptyFolderCandidates(
        dirs: {
          'cloud-top': '/s/Download/Старое',
          'cloud-deep': '/s/Download/Старое/Глубже',
          'cloud-deeper': '/s/Download/Старое/Глубже/Ещё',
        },
        aliveLocally: const {},
        roots: ['/s/Download'],
        rootCloudIds: const {},
      );
      expect(out, [
        '/s/Download/Старое/Глубже/Ещё',
        '/s/Download/Старое/Глубже',
        '/s/Download/Старое',
      ]);
    });
  });
}
